File: Project/gibbsBTM.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <string>
#include <set>
#include <map>
#include <queue>
#include <random>

using namespace std;
int N = 5;
double alpha = 50.0/N;
double beta = 0.01;
int numIter = 100;
int vocabSize;
int numUnits;

map<string,int> wordMap;
map<int,string> wordMapRev;
set<string> vocab;
vector<vector<int> > units;
//topics = []
vector<int> assignments;
int *nz_units;
int *nz_words;
int **nwz;
int **nzw;

void split(const std::string &s, char delim, std::vector<std::string> &elems) {
    std::stringstream ss;
    ss.str(s);
    std::string item;
    while (std::getline(ss, item, delim)) {
        elems.push_back(item);
    }
}

std::vector<std::string> split(const std::string &s, char delim) {
    std::vector<std::string> elems;
    split(s, delim, elems);
    return elems;
}

void readVocab(char* filename){
	int counter = 0;
	ifstream infile(filename);
	string line;
	while (getline(infile,line)){
		vector<string> unit = split(line, ' ');
		for (int i = 0; i < unit.size(); i++){
			vocab.insert(unit[i]);
		}
	}
	set<string>::iterator it;
	for (it = vocab.begin(); it != vocab.end(); it++){
		wordMap.insert(pair<string,int>(*it, counter));
		wordMapRev.insert(pair<int,string>(counter, *it));
		counter++;
	}
}

void readUnits(char* filename){
	ifstream infile(filename);
	string line;
	while (getline(infile,line)){
		vector<string> unit = split(line, ' ');
		vector<int> unit_int;
		for (int i = 0; i < unit.size(); i++){
			unit_int.push_back(wordMap.find(unit[i])->second);
		}
		units.push_back(unit_int);
	}
	numUnits = units.size();
}

void initAssign(char* filename){
	readVocab(filename);
	readUnits(filename);
	vocabSize = vocab.size();
    nwz = new int*[vocabSize];
    nzw = new int*[N];
    nz_words = new int[N];
    nz_units = new int[N];
    for (int i = 0; i < vocabSize; i++){
    	nwz[i] = new int[N];
    }
    for (int i = 0; i < N; i++){
    	nzw[i] = new int[vocabSize];
    }
    for (int u = 0; u < units.size(); u++){
    	int t = rand()%N;
    	assignments.push_back(t);
    	nz_units[t] += 1;
    	nz_words[t] += units[u].size();
    	for (int w = 0; w < units[u].size(); w++){
    		nwz[units[u][w]][t] += 1;
    		nzw[t][units[u][w]] += 1;
    	}
    }
}

void gibbsIteration(){
	std::random_device rd;
    std::mt19937 gen(rd());
    std::vector<double> probs;

	for (int t = 0; t < N; ++t)
		probs.push_back(0.0);
	
	for (int u = 0; u < numUnits; ++u)
	{
		vector<int> unit = units[u];
		int current_topic = assignments[u];
		// double *probs = new double[N];
		
		for (int t = 0; t < N; ++t)
		{
			probs[t] = (nz_units[t]*1.0) + alpha;
			for (int w_iter = 0; w_iter < unit.size(); ++w_iter)
			{
				int w = unit[w_iter];
				probs[t] *= (nwz[w][t]*1.0 + beta)/(nz_words[t] + (vocabSize*beta));
			}
		}
		std::discrete_distribution<> d(probs.begin(), probs.end());
		// Now, sample topic from this distribution.
		int new_topic = d(gen);
		for (int w_iter = 0; w_iter < unit.size(); ++w_iter)
		{
			int w = unit[w_iter];
			nwz[w][new_topic] += 1;
			nzw[new_topic][w] += 1;
			nwz[w][current_topic] -= 1;
			nzw[current_topic][w] -= 1;
		}

		nz_units[current_topic] -= 1;
		nz_units[new_topic] += 1;
		nz_words[current_topic] -= unit.size();
		nz_words[new_topic] += unit.size();
		assignments[u] = new_topic;
	}


}

int main(int argc, char const *argv[])
{


  std::random_device rd;
    std::mt19937 gen(rd());
    std::vector<int> v;
    v.push_back(40);
    v.push_back(10);
    v.push_back(10);
    v.push_back(40);
    std::discrete_distribution<> d(v.begin(), v.end());
    std::map<int, int> m;
    for(int n=0; n<10000; ++n) {
        ++m[d(gen)];
    }
    for(auto p : m) {
        std::cout << p.first << " generated " << p.second << " times\n";
    }

	char* filename = "../train_docs_new.txt";
	initAssign(filename);
	for (int i = 0; i < 500; ++i)
	{
		cout << i <<endl;
		gibbsIteration();
	}

	for (int t = 0; t < N; ++t)
	{
		priority_queue<pair<int,int> > q;
		for (int i = 0; i < vocabSize; ++i)
		{
			q.push(pair<int,int>(nzw[t][i],i));
		}
		for (int i = 0; i < 10; ++i)
		{
			int ki = q.top().second; // index
			cout <<q.top().first << " "<<wordMapRev.find(ki)->second << " ";
			q.pop();
		}
		cout<<endl<<endl;
	}



	return 0;
}
//nwz = {}
//nzw = {}

File: Project/gibbsBTM.py
import sys
import numpy as np
from random import randint

N = 5
alpha = 50.0/N
beta = 0.01
vocab = []
units = []
topics = []
assignments = []
nz_units = []
nz_words = []
nwz = {}
nzw = {}
numIter = 10

def initUnits(unitfile):
	unitFile = open(unitfile)
	for line in unitFile:
		unit = []
		line = line.split("\n")[0]
		line = line.split(" ")
		for w in line:
			if w not in vocab:
				vocab.append(w)
				nwz[w] = [0]*N
			unit.append(w)
		units.append(unit)
	return 0


def initTopics(vocab):
	alpha_array = [alpha]*len(vocab)
	for t in range(0,N):
		topic = np.random.dirichlet(alpha_array)
		topics.append(topic)
		nz_units.append(0)
		nz_words.append(0)
		nzw[t] = [0]*len(vocab)
	return 0

def initAssignments():
	for unit in units:
		topic = randint(0,N-1)
		nz_units[topic] += 1
		nz_words[topic] += len(unit)
		assignments.append(topic)
		for w in unit:
			nwz[w][topic] += 1
			nzw[topic][vocab.index(w)] += 1
	return 0


def gibbsIteration():
	array_topics = []
	for t in range(0,N):
		array_topics.append(t)
	for unit in units:
		current_topic = assignments[units.index(unit)]
		probs = [0.0]*N
		sum_probs = 0.0
		for t in range(0,N):
			probs[t] = (nz_units[t]*1.0) + alpha
			for w in unit:
				probs[t] *= (nwz[w][t]*1.0 + beta)/(nz_words[t] + (len(vocab)*beta))
			sum_probs += probs[t]
		probs_normalized = []
		for p in probs:
			probs_normalized.append(p/sum_probs)
		#print probs_normalized
		new_topic = np.random.choice(array_topics,1,p=list(probs_normalized))[0] 

		for w in unit:
			nwz[w][new_topic] += 1
			nzw[new_topic][vocab.index(w)] += 1
			nwz[w][current_topic] -= 1
			nzw[current_topic][vocab.index(w)] -= 1
		nz_units[current_topic] -= 1
		nz_units[new_topic] += 1
		nz_words[current_topic] -= len(unit)
		nz_words[new_topic] += len(unit)
		assignments[units.index(unit)] = new_topic



initUnits(sys.argv[1])
initTopics(vocab)
print len(vocab)
initAssignments()
for iterations in range(0,numIter):
	print iterations
	gibbsIteration()
for t in range(0,N):
	nzw[t].sort()
	nzw[t].reverse()
	for i in range(10):
		print nzw[t][i]," ",
	print "\n"

